pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/upsertcompactmerge/UpsertCompactMergeTaskGenerator.java

```diff
@@ -0,0 +1,424 @@
+    List<String> segmentsForDeletion = new ArrayList<>();
+    for (String segmentName : validDocIdsMetadataInfoMap.keySet()) {
+      // check if segment is part of completed segments
+      if (!candidateSegmentsMap.containsKey(segmentName)) {
+        LOGGER.warn("Segment {} is not found in the completed segments list, skipping it for {}", segmentName,
+            MinionConstants.UpsertCompactMergeTask.TASK_TYPE);
+        continue;
+      }
+      SegmentZKMetadata segment = candidateSegmentsMap.get(segmentName);
+      for (ValidDocIdsMetadataInfo validDocIdsMetadata : validDocIdsMetadataInfoMap.get(segmentName)) {
+        long totalInvalidDocs = validDocIdsMetadata.getTotalInvalidDocs();
+        long totalValidDocs = validDocIdsMetadata.getTotalValidDocs();
+
+        // Skip segments if the crc from zk metadata and server does not match. They may be getting reloaded.
+        if (segment.getCrc() != Long.parseLong(validDocIdsMetadata.getSegmentCrc())) {
+          LOGGER.warn("CRC mismatch for segment: {}, (segmentZKMetadata={}, validDocIdsMetadata={})", segmentName,
+              segment.getCrc(), validDocIdsMetadata.getSegmentCrc());
+          continue;
+        }
+
+        // segments eligible for deletion with no valid records
+        long totalDocs = validDocIdsMetadata.getTotalDocs();
+        if (totalInvalidDocs == totalDocs) {
+          segmentsForDeletion.add(segmentName);
+        } else if (alreadyMergedSegments.contains(segmentName)) {
+          LOGGER.debug("Segment {} already merged. Skipping it for {}", segmentName,
+              MinionConstants.UpsertCompactMergeTask.TASK_TYPE);
+          continue;
+        } else {
+          Integer partitionID = SegmentUtils.getPartitionIdFromRealtimeSegmentName(segmentName);
+          if (partitionID == null) {
+            LOGGER.warn("Partition ID not found for segment: {}, skipping it for {}", segmentName,
+                MinionConstants.UpsertCompactMergeTask.TASK_TYPE);
+            continue;
+          }
+          segmentsEligibleForCompactMerge.computeIfAbsent(partitionID, k -> new ArrayList<>())
+              .add(new SegmentMergerMetadata(segment, totalValidDocs, totalInvalidDocs));
+        }
+        break;
+      }
+    }
+
+    segmentsEligibleForCompactMerge.forEach((partitionID, segmentList) ->
+        segmentList.sort(Comparator.comparingLong(o -> o.getSegmentZKMetadata().getCreationTime()))
+    );
+
+    // Map to store the result: each key (partition) will have a list of groups
+    Map<Integer, List<List<SegmentMergerMetadata>>> groupedSegments = new HashMap<>();
+
+    // Iterate over each partition and process its segments list
+    for (Map.Entry<Integer, List<SegmentMergerMetadata>> entry : segmentsEligibleForCompactMerge.entrySet()) {
+      int partitionID = entry.getKey();
+      List<SegmentMergerMetadata> segments = entry.getValue();
+      // task config thresholds
+      // TODO add output segment size as one of the thresholds
```

Comment:
Can we create a issue for this for tracking?

Reply:
https://github.com/apache/pinot/issues/14634